LFS-1194: As a user, I can filter forms by "Created date" in the dashboard and on the Forms page

IntegralProgrammer opened the pull request:
This PR introduces a _Created Date_ filter for selecting Forms that are `=`, `<>`, `<`, `>`, `<=`, or `>=` a given date.

To test:

1. Build this branch (`git checkout LFS-1194`, `mvn clean install`)
2. Start CARDS with the `lfs` runMode enabled
3. Create a new Form (and associated Subject). Fill in all mandatory questions so that the Form is marked as _complete_
4. Go to the Dashboard
5. Filter the completed Forms by _Created Date_. Experiment with various filters (eg. Filtering for `Created Date` `=` _date of Form creation_ should list the Form, while filtering for `Created Date` `<>` _date of Form creation_ should _not_ list the Form)
6. Go to the _Forms_ page and perform the same tests as done in step 5

IntegralProgrammer committed:
LFS-1194: As a user, I can filter forms by "Created date" in the dashboard and on the Forms page

IntegralProgrammer commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/PaginationServlet.java:
```diff
@@ -278,6 +283,67 @@ private String createSingleJoin(final String[] joins, final String childprefix,
         return joindata.toString();
     }
 
+    private String generateDateCompareQuery(String jcrVariable, String thisDayStr, String operator)
```
This method is useful for checking if the (Year, Month, Date) component of any JCR property of type `Date` is `=`, `<>`, `<`, `>`, `<=`, or `>=` a date string specified in the `YYYY-MM-DD` format. Should it be moved to somewhere else it the repository where it could potentially be used by other functions?

IntegralProgrammer commented:
There is also the issue of _timezone in the client (browser)_ vs. _timezone in the server_. Running CARDS on the host will use the host's timezone but running CARDS in a Docker container will use the UTC timezone.

Should therefore, the client's timezone be sent along with the filter `fetch` request so that only Forms that match the _Created Date_ criteria based on _their_ local timezone will be selected?

sashaandjic commented:
Looks good and works well. Once small suggestion: can we add some space between calendar and `X`.

![image](https://user-images.githubusercontent.com/23243244/122353641-3b8d0c00-cf1e-11eb-84db-ce6b95d0e41b.png)

IntegralProgrammer committed:
LFS-1194: As a user, I can filter forms by "Created date" in the dashboard and on the Forms page

Pass the client's timezone to the PaginationServlet so that date of creation filtering will be done relative to the client's timezone

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -239,6 +277,7 @@ function Filters(props) {
         toCheck
         :
         {...toCheck, comparator: (toCheck.comparator == "=" ? "is empty" : "is not empty")}));
+    newFilters = addCreatedDateTimezone(newFilters);
```
I'm wondering if there's a better location for this code to be placed. What needs to be done for `"createddate"` filters is that the client's timezone needs to be added when the value is sent to the server but it needs to be removed when displayed in the filter editing dialog otherwise it is shown (and later changed to) the blank template `mm/dd/yyyy`.

IntegralProgrammer commented:
Also, currently the _active filter_ Material-UI chips displays the _Created Date_ filter with the associated timezone. Personally, I like this but I'm wondering what would be best for our users?

![cards_filters](https://user-images.githubusercontent.com/5349728/122592137-ecd09680-d031-11eb-80b7-eaafd659cd5f.png)

sashaandjic commented:
> Also, currently the _active filter_ Material-UI chips displays the _Created Date_ filter with the associated timezone. Personally, I like this but I'm wondering what would be best for our users?

I would go without time zone.

IntegralProgrammer committed:
LFS-1194: As a user, I can filter forms by "Created date" in the dashboard and on the Forms page

Do not display the timezone in the "Created date" filter Material-UI <Chip>

IntegralProgrammer commented:
> > Also, currently the _active filter_ Material-UI chips displays the _Created Date_ filter with the associated timezone. Personally, I like this but I'm wondering what would be best for our users?
> 
> I would go without time zone.

Implemented by commit `68d48c6` :heavy_check_mark:

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (commented):
This looks good so far, and I'm not worried yet about finding the perfect place for the newly added methods. However, what bothers me is that we already have support for date filters (for actual date questions), yet we needed a lot of extra code for the creation date. Is the code we had good enough for created date as well, or should the new code be also used for regular date questions?

IntegralProgrammer commented:
> This looks good so far, and I'm not worried yet about finding the perfect place for the newly added methods. However, what bothers me is that we already have support for date filters (for actual date questions), yet we needed a lot of extra code for the creation date. Is the code we had good enough for created date as well, or should the new code be also used for regular date questions?

The code that we had for the other types of _date_ filtering, (eg. _Date of birth_, _Date of genetic diagnosis_, etc...) doesn't work for filtering by _created date_. The reason for that is because when a date is _entered_ such as _Date of birth_ or _Date of genetic diagnosis_ the user (though the front-end) controls the value that is sent to the server and as a _time_ is not specified with the sent date, a default time of `00:00:00.000Z` is assumed. Under that assumption, if a _Date of birth_ were to be, for example, `2021-07-01` it would be stored in JCR as `2021-07-01T00:00:00.000` and thus checking for equality with `2021-07-01` makes sense as it gets converted to `2021-07-01T00:00:00.000` and the comparison to `2021-07-01T00:00:00.000` evaluates to `True`.

The issue with `jcr:created` is that the date/time stored in JCR holds the time that the Form node was created at. So for example, if a Form were to be created at `2021-06-25T10:32:45.431`, using the aforementioned method for querying for Forms created on `2021-06-25` would fail as `2021-06-25` would be converted to `2021-06-25T00:00:00.000` which does not equal `2021-06-25T10:32:45.431`.

The new code _could_, but doesn't have to, be used with manually entered date questions (eg. `Date of birth`) as being at a time of `00:00:00.000` with _Date X_  is really just a special case of _Date X_. Nonetheless, should using this new code with manually entered dates be done just for consistency?
